// contour-framework/ui/control/heading-bar.reel/heading-bar.js
/**
 * @module ui/control/heading-bar.reel
 * @requires montage/ui/component
 */
var Component = require("montage/ui/component").Component;

/**
 * @class HeadingBar
 * @extends Component
 */
exports.HeadingBar = Component.specialize(/** @lends HeadingBar.prototype */ {

    constructor: {
        value: function HeadingBar() {

        }
    },

    heading: {
        value: null
    },

    level: {
        value: 1
    },

    _hasAction: {
        value: false
    },

    _actionLabel: {
        value: null
    },

    actionLabel: {
        set: function(label) {
            this._hasAction = (label !== null && typeof label !== 'undefined');
            this._actionLabel = label;
        },
        get: function() {
            return this._actionLabel;
        }
    },

    _preparedForAction: {
        value: false
    },

    prepareForActivationEvents: {
        value: function () {
            if (!this._preparedForAction) {
                this.addEventListener("action", this, false);
                this._preparedForAction = true;
            }
        }
    },

    handleButtonAction: {
        value: function (event) {
            event.stopPropagation();
            this.dispatchEventNamed("headingAction", true, true, {headingBar: this});
        }
    }

});
