// contour-framework/logic/layer/wms-loader.js
var Target = require("montage/core/target").Target;
var Request = require("montage/core/request");
var Q = require("q").Q;

var WmsLoader = exports.WmsLoader = Target.specialize({ }, {
    load: {
        value: function( url ) {
            url = WmsLoader.prepareUrl(url);

            return Request.request(url, {
                withCredentials: false
            }).then(function( resp ) {
                if ( resp.status !== 200 ) {
                    return Q.reject();
                }
                return WmsLoader.parseWmsService(url, resp.body);
            });
        }
    },
    parseWmsService: {
        value: function( url, xml ) {
            var service;
            var parser = new DOMParser();
            var xmlDoc = parser.parseFromString(xml, "text/xml");

            var $service = xmlDoc.getElementsByTagName("Service")[0];

            var $layers = [].slice.call(xmlDoc.querySelectorAll("Capability > Layer > Layer"));
            var name = WmsLoader.getText($service.getElementsByTagName("Name")[0]);
            var title = WmsLoader.getText($service.getElementsByTagName("Title")[0]);
            var layers = $layers.map(function( layer ) {
                return WmsLoader.parseLayer(layer);
            });

            service = {
                name: name.replace(/_/g, " "),
                title: title.replace(/_/g, " "),
                url: url.split("?")[0] + "?",
                description: WmsLoader.getText($service.getElementsByTagName("Abstract")[0]),
                protocol: "WMS",
                coordSysId: "web mercator",
                protocolVersion: this.getVersion(xmlDoc),
                layers: layers
            };

            return {
                service: service,
                type: "service"
            };
        }
    },
    parseLayer: {
        value: function( layer ) {
            var title = WmsLoader.getText(layer.getElementsByTagName("Title")[0]);
            var name = WmsLoader.getText(layer.getElementsByTagName("Name")[0]);
            var $legend = layer.getElementsByTagName("LegendURL")[0];
            var $swatch = $legend ? $legend.getElementsByTagName("OnlineResource")[0] : "";
            var swatchUrl;

            if ( $swatch ) {
                swatchUrl = $swatch.getAttribute("xlink:href");
            }

            return {
                title: title.replace(/_/g, " "),
                name: name.replace(/_/g, " "),
                type: "WmsRasterLayer",
                coordSysId: this.getCoordSysId(WmsLoader.getText(layer.getElementsByTagName("SRS")[0])),
                description: WmsLoader.getText(layer.getElementsByTagName("Abstract")[0]),
                swatch: swatchUrl
            };
        }
    },
    getText: {
        value: function( node ) {
            if ( !node ) {
                return "";
            }

            var children = node.childNodes;
            var cdataReg = /<!\[CDATA\[(.*)]]>/;
            var text, match;

            if ( children.length ) {
                return this.getText(children[0]);
            }

            text = node.textContent;
            match = cdataReg.exec(text);

            if ( match ) {
                text = match[1];
            }

            return text;
        }
    },
    getVersion: {
        value: function( kmlDoc ) {
            var $wmsService = kmlDoc.children[0];
            return $wmsService.getAttribute("version");
        }
    },
    getCoordSysId: {
        value: function( crs ) {
            //var config = Config.ServerConfig.getInstance();
            //var coordSystems = config.coordSystems.items;
            //var coordSys;
            //var arr = crs.split(":");
            //crs = arr[1];

            //if ( crs ) {
            //    coordSys = _.find(coordSystems, { epsgId: crs });
            //    if ( coordSys ) {
            //        return coordSys.id;
            //    }
            //}
            return crs || "web mercator";
        }
    },
    prepareUrl: {
        value: function( url ) {
            var arr = url.split("?");
            return arr[0] + "?service=WMS&request=GetCapabilities";
        }
    }
});
