// contour-framework/ui/brief/coords/brief-circle.reel/brief-circle.js
/**
 * @module ui/brief-circle.reel
 */
var Component = require("montage/ui/component").Component;

/**
 * @class BriefCircle
 * @extends Component
 */
exports.BriefCircle = Component.specialize(/** @lends BriefCircle# */ {
    constructor: {
        value: function BriefCircle() {

        }
    },

    _area: { value: null },
    coords: { value: null },

    area: {
        set: function(value) {
            this._area = value;
            if (value && value.bbox) {
                this._buildStringifyCoordinates(value);
            }
        },

        get: function() {
            return this._area;
        }
    },

    _buildStringifyCoordinates: {
        value: function(value) {
            var buff = (value.bbox[3] - value.bbox[1]) / 2;
            var lat = value.bbox[1] + buff;
            var lon = value.bbox[0] + buff;
            this.coords = {
                lat: Math.abs(lat).toFixed(2) + '\u00B0 ' + (lat < 0 ? 'S' : 'N'),
                lon: Math.abs(lon).toFixed(2) + '\u00B0 ' + (lon < 0 ? 'W' : 'E'),
                radius: Math.abs(buff).toFixed(2) + '\u00B0'
            };
        }
    }
});
