// contour-framework/ui/layer/layer-list.reel/layer-list.js
var Component = require("montage/ui/component").Component,
    PressComposer = require("montage/composer/press-composer").PressComposer;

/**
 * @class LayerList
 * @extends external:Component
 */
exports.LayerList = Component.specialize(/** @lends LayerList.prototype */{

    // Set by owner.
    layers: {
        value: undefined
    },

    // Set by owner.
    hasSettings: {
        value: false
    },

    // Set by owner.
    succession: {
        value: undefined
    },

    // Set in serialization.
    _list: {
        value: undefined
    },

    _settings: {
        value: {
            isCustomizing: false,
            sortOrder: "A-Z",
            areEnabledLayersFirst: false
        }
    },

    prepareForActivationEvents: {
        value: function () {
            this._pressComposer = new PressComposer();
            this.addComposerForElement(this._pressComposer, this._list.element);
            this._pressComposer.addEventListener("press", this);

            this.addEventListener("action", this, false); // for drillDownButton
        }
    },

    handlePress: {
        value: function (event) {
            event.stopPropagation();
            if (event.targetElement.component) {
                event.targetElement.component.checkbox.checked = !event.targetElement.component.checkbox.checked;
            }
        }
    },

    handleDrillDownButtonAction: {
        value: function (event) {
            event.stopPropagation();
            this.application.delegate.inspectLayer(event.target.ownerComponent.node.data);
        }
    },

    // Used in serialization.
    _getTitle: {
        value: function (layers) {
            var count = layers ? layers.length : -1;
            return count === 1 ? "1 Layer" :
                   count >= 0 ?   count + " Layers" :
                                  "";
        }
    }

});
