// contour-framework/ui/product/product-list-item.reel/product-list-item.js
/**
 * @module "ui/product/product-list-item.reel"
 * @requires montage/ui/component
 */
var Component = require("montage/ui/component").Component,
    MapDataConverter = require("logic/map/converters/map-data-converter").MapDataConverter,
    ProjectionUtilities = require("logic/map/projection-utilities").ProjectionUtilities,
    Projections = require("logic/map/projection-utilities").Projections,
    PressComposer = require("montage/composer/press-composer").PressComposer;

/**
 * @class ProductListItem
 * @extends Component
 */
var ProductListItem = exports.ProductListItem = Component.specialize(/** @lends ProductListItem.prototype */{

    _isImageFile: {
        value: function () {
            return this._fileExtension && ProductListItem._imageExtensions.hasOwnProperty(this._fileExtension);
        }
    },

    __fileExtension: { value: null },

    _fileExtension: {
        get: function () {
            if (this._product && this._product.type && this._product.type.id === "FILE_PRODUCT") {
                var lastDot = this._product.fileName.lastIndexOf(".");
                this.__fileExtension = this._product.fileName.substring(lastDot+1).trim().toUpperCase();

            }
            return this.__fileExtension;
        }
    },

    _dateConverter: {
        value: {
            convert: function(input) {
                return input ? input.format("M.DD.YYYY H\\h:mm\\m") : "";
            }
        }
    },

    _product: { value: null },

    _productDate: {
        set: function (value) {
            if (value) {
                value.converter = this._dateConverter;
            }
        }
    },

    //__contentElement: { value: null },
    //
    //_contentElement: {
    //    set: function (value) {
    //        this.__contentElement = value;
    //        this._assignUrlToTitleElement();
    //    }
    //},

    //_assignUrlToTitleElement: {
    //    value: function () {
    //        var url;
    //        if (this._product && this.__contentElement) {
    //            if (this._product.type === "URL_PRODUCT") {
    //                url = this._product.url;
    //            } else if (this._product.type === "FILE_PRODUCT" || this._product.type === "TEXT_PRODUCT") {
    //                url = this._product.previewUrl;
    //            }
    //            this.__contentElement.setAttribute("href", url);
    //        }
    //    }
    //},

    latitudeString: {
        get: function () {
            var string;
            if (!this.product || !this.product.coordinates) {
                string = "";
            } else if (this.product.coordinates[1] < 0) {
                string = (this.product.coordinates[1]*-1).toFixed(2) + " S"
            } else {
                string = (this.product.coordinates[1]).toFixed(2) + " N"
            }
            return string;
        }
    },

    longitudeString: {
        get: function () {
            var string;
            if (!this.product || !this.product.coordinates) {
                string = "";
            } else if (this.product.coordinates[0] < 0) {
                string = (this.product.coordinates[0]*-1).toFixed(2) + " E"
            } else {
                string = (this.product.coordinates[0]).toFixed(2) + " W"
            }
            return string;
        }
    },

    coordinateString: {
        get: function () {
            if (!this._coordinateString) {
                this._coordinateString = this.latitudeString + '\u00B0, ' + this.longitudeString  + '\u00B0';
            }
            return this._coordinateString;
        },
        set: function (value) {
            if (value !== this._coordinateString) {
                this._coordinateString = value;
            }
        }
    },

    product: {
        get: function () {
            return this._product;
        },
        set: function (value) {
            if (this._product !== value) {
                this._product = value;
                this._fileExtension = null;
                this.coordinateString = null;
            }
        }
    }


}, {

    _imageExtensions: {
        value: {
            JPEG: null,
            JPG: null,
            PNG: null,
            GIF: null
        }
    }

});
