// contour-framework/ui/brief/brief-inspector-details.reel/brief-inspector-details.js
/**
 * @module "ui/brief-inspector-details.reel"
 */
var Component = require("montage/ui/component").Component;

/**
 * @class BriefInspectorDetails
 * @extends Component
 */
exports.BriefInspectorDetails = Component.specialize(/** @lends BriefInspectorDetails.prototype */{
    tab: {
        value: null
    },

    url: {
        value: ""
    },

    _briefDetails: {
        value: null
    },

    briefDetails: {
        get: function () {
            return this._briefDetails;
        },
        set: function (value) {
            if (!value) {
                return;
            }

            this._briefDetails = value;

            if (this.tab && this.tab._index !== -1) {
                this.url = this.getUrl();
            } else {
                this.url = "";
            }

            this.needsDraw = true;
        }
    },

    getUrl: {
        value: function () {
            if (!this.briefDetails) return "";

            var templateId = this.briefDetails.brief.templateId,
                tabId = this.briefDetails.brief.tabs[this.tab._index].id,
                bbox = this.briefDetails.area.bbox,
                figureType = this.briefDetails.area.type;

            if (figureType === "Rectangle") {
                return "http://atlastest.pdc.org/atlas/area_brief_proxy/ab_proxy.jsp?" +
                    "http://servicestest.pdc.org/control/atlas/rr.php?shapetype=rect" +
                    "&xmin=" + bbox[0] + "&ymin=" + bbox[1] + "&xmax=" + bbox[2] + "&ymax=" +
                    bbox[3] + "&tab_id=" + tabId + "&tpl_id=" + templateId + "&version=2&async=1";
            }

            if (figureType === "Circle") {
                var buff = ((bbox[3] - bbox[1]) / 2);
                return "http://atlastest.pdc.org/atlas/area_brief_proxy/ab_proxy.jsp?" +
                    "http://servicestest.pdc.org/control/atlas/rr.php?shapetype=circ" +
                    "&x=" + (bbox[0] + buff) + "&y=" + (bbox[1] + buff) + "&buff=" + buff +
                    "&tab_id=" + tabId + "&tpl_id=" + templateId + "&version=2&async=1";
            }
        }
    }
});
