// contour-framework/ui/bookmark/bookmark-list-settings.reel/bookmark-list-settings.js
/**
 * @module ui/bookmark-list-settings.reel
 */
var Component = require("montage/ui/component").Component;

/**
 * @class BookmarkListSettings
 * @extends Component
 */
exports.BookmarkListSettings = Component.specialize(/** @lends BookmarkListSettings.prototype */ {
    enterDocument: {
        value: function (firstTime) {
            if (!firstTime) {
                this.prepareForActivationEvents();
            }
        }
    },
    exitDocument: {
        value: function () {
            if (this._preparedForAction) {
                this._preparedForAction = false;
                this.templateObjects.title.removeEventListener("backAction", this);
                this.templateObjects.sortLegendBar.removeEventListener("action", this);
            }
        }
    },
    handleBackAction: {
        value: function (e) {
            e.stopPropagation();
            this.dispatchEventNamed("exitBookmarkListSettingsAction", false, false, {});
        }
    },
    handleSortLegendBarAction: {
        value: function (e) {
            e.stopPropagation();
            this.templateObjects.sortFieldSelectBar.value = this.templateObjects.sortFieldSelectBar.default;
            this.templateObjects.sortDirectionSelectBar.value = this.templateObjects.sortDirectionSelectBar.default;
        }
    },
    _preparedForAction: {
        value: false
    },
    prepareForActivationEvents: {
        value: function () {
            if (!this._preparedForAction) {
                this._preparedForAction = true;
                this.templateObjects.title.addEventListener("backAction", this, false);
                this.templateObjects.sortLegendBar.addEventListener("action", this, false);
            }
        }
    }
});
