// contour-framework/ui/layer/import/layer-user-wms-service-inspector.reel/layer-user-wms-service-inspector.js
/**
 * @module ui/layer/layer-import-file-inspector.reel
 * @requires montage/ui/component
 */
var Component = require("montage/ui/component").Component,
    WmsLoader = require("logic/layer/wms-loader").WmsLoader,
    UserLayersService = require("logic/mock-service/user-layers-service").UserLayersService;

/**
 * @class LayerUserWmsServiceInspector
 * @extends Component
 */
exports.LayerUserWmsServiceInspector = Component.specialize(/** @lends LayerUserWmsServiceInspector# */ {
    errorMessage: {
        value: ""
    },

    service: {
        value: null
    },

    constructor: {
        value: function LayerImportFileInspector() {

        }
    },

    enterDocument: {
        value: function () {
            this.serviceUrl = "";
            this.service = null;
        }
    },
    handleCancelButtonAction: {
        value: function (event) {
            event.stopPropagation();
            this.service = null;
        }
    },
    handleAction: {
        value: function handleImportButtonAction() {
            var self = this,
                service = UserLayersService.getServiceByUrl(this.serviceUrl),
                serviceUrl = this.serviceUrl;

            if (service) {
                self.dispatchEventNamed("serviceLoaded", true, true, service);
            } else {
                this.service = {
                    url: serviceUrl
                };

                self.errorMessage = "";

                WmsLoader.load(this.serviceUrl)
                    .then(function (data) {
                        if (self.service && self.service.url === serviceUrl) {
                            service = UserLayersService.addService(data.service);
                            self.dispatchEventNamed("serviceLoaded", true, true, service);
                        }
                    }, function (error) {
                        if (self.service && self.service.url === serviceUrl) {
                            self.errorMessage = error;
                            self.service = null;
                        }
                    });
            }
        }
    }
});
