// contour-framework/ui/link/link-list.reel/link-list.js
/**
 * @module ui/link/link-list.reel
 */
var Component = require("montage/ui/component").Component,
    PressComposer = require("montage/composer/press-composer").PressComposer;

/**
 * @class LinkList
 * @extends Component
 */
exports.LinkList = Component.specialize(/** @lends LinkList# */ {
    constructor: {
        value: function LinkList() {

        }
    },
    _links: {
        value: []
    },
    links: {
        set: function (value) {
            if (value && Array.isArray(value) && value !== this._links) {
                this._links = value;
            }
        },
        get: function() {
            return this._links;
        }
    },
    handlePress: {
        value: function (event) {
            event.stopPropagation();
            var component = event.targetElement.component;
            if (component && component.link && component.link.url) {
                window.open(component.link.url, '_blank').focus();
            }
        }
    },

    enterDocument: {
        value: function (firstTime) {
            if (!firstTime) {
                this.prepareForActivationEvents();
            }
        }
    },
    exitDocument: {
        value: function () {
            if (this._preparedForAction) {
                this._preparedForAction = false;
                this._pressComposer.removeEventListener("press", this);
                this._pressComposer = null;
            }
        }
    },
    _preparedForAction: {
        value: false
    },
    prepareForActivationEvents: {
        value: function () {
            if (!this._preparedForAction) {
                this._preparedForAction = true;
                this._pressComposer.addEventListener("press", this, false);
            }
        }
    },
    __pressComposer: {
        value: null
    },
    _pressComposer: {
        get: function () {
            if (!this.__pressComposer) {
                this.__pressComposer = new PressComposer();
                this.addComposer(this.__pressComposer);
            }
            return this.__pressComposer;
        },
        set: function (value) {
            if (value !== this.__pressComposer) {
                if (this.__pressComposer) {
                    this.removeComposer(this.__pressComposer);
                }
                this.__pressComposer = value;
                if (value) {
                    this.addComposer(this.__pressComposer);
                }
            }
        }
    }
});
