// contour-framework/ui/camera/camera-ozolio-viewer.reel/camera-ozolio-viewer.js
/**
 * @module ui/camera/camera-ozolio-viewer.reel
 */
var Component = require("montage/ui/component").Component;

/**
 * @class CameraOzolioViewer
 * @extends Component
 */
exports.CameraOzolioViewer = Component.specialize(/** @lends CameraOzolioViewer# */ {
    constructor: {
        value: function CameraOzolioViewer() {

        }
    },

    _camera: { value: null },

    _areScriptsLoaded: {
        get: function () {
            return this._numScriptsLoaded === 3;
        }
    },

    _numScriptsLoaded: { value: 0 },

    _onScriptLoad: {
        value: function () {
            this._numScriptsLoaded++;
            if (this._areScriptsLoaded && this._camera) {
                this._isNewCamera = true;
                this.needsDraw = true;
            }
        }
    },

    _ozolioScript: {
        set: function (value) {
            if (value) {
                value.onload = this._onScriptLoad.bind(this);
            }
        }
    },

    _prepareWebcam: {
        value: function () {
            var camvars = {
                    auto_play: "true",
                    domain: "https://relay.ozolio.com",
                    camera_doc: this._camera.properties.camId
                },
                params = {
                    allowFullScreen: "true",
                    allowScriptAccess: "always"
                };
            this._webcam = new jwebcam(camvars, "camera_"+this._camera.id, 420, 360, params);
        }
    },

    _swfScript: {
        set: function (value) {
            if (value) {
                value.onload = this._onScriptLoad.bind(this);
            }
        }
    },

    _viewerElement: { value: null },

    _webcam: { value: null },

    _webcamScript: {
        set: function (value) {
            if (value) {
                value.onload = this._onScriptLoad.bind(this);
            }
        }
    },

    camera: {
        get: function () {
            return this._camera;
        },
        set: function (value) {
            if (this._camera !== value) {
                if (this._camera) {
                    //do something
                }
                this._camera = value;
                if (this._camera && this._areScriptsLoaded) {
                    this._isNewCamera = true;
                    this.needsDraw = true;
                }
            }
        }
    },

    draw: {
        value: function () {
            if (this._camera) {
                this._viewerElement.setAttribute("id", "camera_"+this._camera.id);
                if (this._isNewCamera) {
                    this._isNewCamera = false;
                    this._prepareWebcam();
                }
            }
        }
    },

    exitDocument: {
        value: function () {
            if (this._webcam) {
                this._webcam.exit();
                this._webcam = null;
            }
        }
    }



});
