// contour-framework/ui/social/social-photo.reel/social-photo.js
/**
 * @module ui/social-photo.reel
 */
var Component = require("montage/ui/component").Component;

/**
 * @class SocialPhoto
 * @extends Component
 */
exports.SocialPhoto = Component.specialize(/** @lends SocialPhoto# */ {

    constructor: {
        value: function SocialPhoto() {

        }
    },

    _preparedForAction: { value: false },

    _settingsPane: {value: null},

    enterDocument: {
        value: function (firstTime) {
            if (!firstTime) {
                this.prepareForActivationEvents();
            }
        }
    },

    exitDocument: {
        value: function () {
            if (this._preparedForAction) {
                this._preparedForAction = false;
                this._titleBar.removeEventListener("settingsAction", this, false);
            }
        }
    },

    handleSettingsAction: {
        value: function (e) {
            e.stopPropagation();
            this.settingsPane.succession = this.succession;
            this.succession.components.push(this.settingsPane);
        }
    },

    prepareForActivationEvents: {
        value: function () {
            if (!this._preparedForAction) {
                this._preparedForAction = true;
                this._titleBar.addEventListener("settingsAction", this, false);

            }

        }
    },

    settingsPane: { value: null },

    succession: { value: null }

});
