// contour-framework/ui/control/status-bar.reel/status-bar.js
/**
 * @module ui/control/status-bar.reel
 * @requires montage/ui/component
 */
var Component = require("montage/ui/component").Component,
    MutableEvent = require("montage/core/event/mutable-event").MutableEvent;

/**
 * @class StatusBar
 * @extends Component
 */
exports.StatusBar = Component.specialize(/** @lends StatusBar.prototype */ {

    hasCancel: {
        value: true
    },

    title: {
        value: null
    },

    isVisible: {
        value: false
    },

    constructor: {
        value: function StatusBar() {

        }
    },

    /***************************************************************************
     * Button events are handled here and converted to title bar events.
     */
    _preparedForAction: {
        value: false
    },

    prepareForActivationEvents: {
        value: function () {
            if (!this._preparedForAction) {
                this.addEventListener("action", this, false);
                this._preparedForAction = true;
            }
        }
    },

    handleCancelAction: {
        value: function () {
            this.isVisible = false;
        }
    }
});
