// contour-framework/ui/brief/coords/brief-rectangle.reel/brief-rectangle.js
/**
 * @module ui/brief-rectangle.reel
 */
var Component = require("montage/ui/component").Component;

/**
 * @class BriefRectangle
 * @extends Component
 */
exports.BriefRectangle = Component.specialize(/** @lends BriefRectangle# */ {
    constructor: {
        value: function BriefRectangle() {

        }
    },

    _area: { value: null },
    coords: { value: null },

    area: {
        set: function(value) {
            this._area = value;
            if (value && value.bbox) {
                this._buildStringifyCoordinates(value);
            }
        },

        get: function() {
            return this._area;
        }
    },

    _buildStringifyCoordinates: {
        value: function(value) {
            this.coords = {
                south: Math.abs(value.bbox[1]).toFixed(2) + '\u00B0 ' + (value.bbox[1] < 0 ? 'S' : 'N'),
                north: Math.abs(value.bbox[3]).toFixed(2) + '\u00B0 ' + (value.bbox[3] < 0 ? 'S' : 'N'),
                west: Math.abs(value.bbox[0]).toFixed(2) + '\u00B0 ' + (value.bbox[0] < 0 ? 'W' : 'E'),
                east: Math.abs(value.bbox[2]).toFixed(2) + '\u00B0 ' + (value.bbox[2] < 0 ? 'W' : 'E')
            };
        }
    }
});
