// contour-framework/ui/control/edit-bar.reel/edit-bar.js
/**
 * @module ui/control/edit-bar.reel
 * @requires montage/ui/component
 */
var Component = require("montage/ui/component").Component,
    MutableEvent = require("montage/core/event/mutable-event").MutableEvent;

/**
 * @class EditBar
 * @extends Component
 */
exports.EditBar = Component.specialize(/** @lends EditBar.prototype */ {
    constructor: {
        value: function EditBar() {

        }
    },

    /***************************************************************************
     * Properties
     */
    hasAdd: {
        value: false
    },

    hasCancel: {
        value: false
    },

    hasCreate: {
        value: false
    },

    hasImport: {
        value: false
    },

    hasSave: {
        value: false
    },

    title: {
        value: null
    },

    /***************************************************************************
     * Button events are handled here and converted to title bar events.
     */
    _preparedForAction: {
        value: false
    },

    prepareForActivationEvents: {
        value: function () {
            if (!this._preparedForAction) {
                this.addEventListener("action", this, false);
                this._preparedForAction = true;
            }
        }
    },

    handleAddAction: {
        value: function (event) {
            event.stopPropagation();
            this.dispatchEventNamed("addAction", true, true, {});
        }
    },

    handleCreateAction: {
        value: function (event) {
            event.stopPropagation();
            this.dispatchEventNamed("createAction", true, true, {});
        }
    },

    handleCancelAction: {
        value: function (event) {
            event.stopPropagation();
            this.dispatchEventNamed("cancelAction", true, true, {});
        }
    },

    handleClearAction: {
        value: function (event) {
            event.stopPropagation();
            this.dispatchEventNamed("clearAction", true, true, {});
        }
    },


    handleImportAction: {
        value: function (event) {
            event.stopPropagation();
            this.dispatchEventNamed("importAction", true, true, {});
        }
    },

    handleSaveAction: {
        value: function (event) {
            event.stopPropagation();
            this.dispatchEventNamed("saveAction", true, true, {});
        }
    },

    handleUndoAction: {
        value: function (event) {
            event.stopPropagation();
            this.dispatchEventNamed("undoAction", true, true, {});
        }
    }
});
