// contour-framework/ui/control/coordinate-editor.info/sample/ui/main.reel/main.js
/**
 * @module ui/main.reel
 */
var Component = require("montage/ui/component").Component;

/**
 * @class Main
 * @extends Component
 */
exports.Main = Component.specialize(/** @lends Main# */ {
    constructor: {
        value: function Main() {

        }
    },

    _googleMarker: {
        value: null
    },

    _latitude: {
        value: 67
    },

    latitude: {
        get: function () {
            return this._latitude;
        },
        set: function (value) {
            if (this._latitude !== value) {
                this._latitude = value;
                this._updateMarker();
            }
        }
    },

    longitude: {
        get: function () {
            return this._longitude;
        },
        set: function (value) {
            if (this._longitude !== value) {
                this._longitude = value;
                this._updateMarker();
            }
        }
    },

    _updateMarker: {
        value: function () {
            if (this._googleMarker) {
                this._googleMarker.setPosition(new google.maps.LatLng(this._latitude, this._longitude));
            }

        }
    },


    _longitude: {
        value: -153
    },



    coordinate: {
        value: {
            latitude: 67,
            longitude: -153
        }
    },

    _preparedForAction: {
        value: false
    },

    prepareForActivationEvents: {
        value: function () {
            if (!this._preparedForAction) {
                this._preparedForAction = true;
                this.coordinateEditor.addEventListener("coordinateAction", this, false);
                var coord = this.coordinate,
                    self = this;
                this._googleMarker = new google.maps.Marker({
                    position: new google.maps.LatLng(coord.latitude, coord.longitude),
                    map: self.map.mapEngine.map
                });
            }
        }
    },

    _selectingLocation: {
        value: false
    },

    handleCoordinateAction: {
        value: function (e) {
            e.stopPropagation();
            if (this._selectingLocation) {
                this._selectingLocation = false;
                this.coordinateEditor.headerActionLabel = "Select Location";
                this.map.removeEventListener("mapClickAction",this);
                this.map.mapEngine.map.setOptions({draggableCursor: null});
            } else {
                this._selectingLocation = true;
                this.coordinateEditor.headerActionLabel = "Stop Selecting";
                this.map.addEventListener("mapClickAction", this, false);
                this.map.mapEngine.map.setOptions({draggableCursor: "crosshair"});
            }


        }
    },

    handleMapClickAction: {
        value: function (e) {
            e.stopPropagation();
            this.longitude = e.detail.coordinate[0],
            this.latitude = e.detail.coordinate[1];
        }
    }

});
