// contour-framework/logic/mock-model/area-brief.js
var Montage = require("montage").Montage,
    GeometryUtilities = require("logic/map/geometry-utilities").GeometryUtilities;

var AreaBrief = exports.AreaBrief = Montage.specialize(/** @lends AreaBrief.prototype */ {

    bbox: { value: null },

    center: { value: null},

    radius: { value: null},

    type: { value: null },

    name: { value: null }

}, {

    initializeAreaBrief: {
        value: function (name, boundingBox, type) {
            var areaBrief = new AreaBrief();
            areaBrief.name = name;
            areaBrief.bbox = boundingBox;
            areaBrief.type = type;
            if (type === "Circle") {
                var radius = Math.abs(boundingBox[0] - boundingBox[3]),
                    centerLong = boundingBox[0] + radius;
                if (centerLong > 180) {
                    centerLong = -360 + centerLong;
                }
                areaBrief.center = [centerLong, boundingBox[1]+radius];
                areaBrief.radius = radius;
            }
            return areaBrief;
        }
    }

});
