// contour-framework/ui/link/link-category-list.reel/link-category-list.js
/**
 * @module ui/link/link-category-list.reel
 */
var Component = require("montage/ui/component").Component;

/**
 * @class LinkList
 * @extends Component
 */
exports.LinkCategoryList = Component.specialize(/** @lends LinkCategoryList# */ {
    constructor: {
        value: function LinkCategoryList() {

        }
    },
    _categories: {
        value: []
    },
    categories: {
        set: function (value) {
            if (value && Array.isArray(value) && value !== this._categories) {
                this._categories = value;
            }
        },
        get: function() {
            return this._categories;
        }
    },

    handleSelectCategoryLinkListItemAction: {
        value: function(event) {
            event.stopPropagation();
            this.dispatchEventNamed("selectLinkCategory", true, true, event.detail);
        }
    },
    enterDocument: {
        value: function (firstTime) {
            if (!firstTime) {
                this.prepareForActivationEvents();
            }
        }
    },
    _preparedForAction: {
        value: false
    },
    prepareForActivationEvents: {
        value: function () {
            if (!this._preparedForAction) {
                this.addEventListener("selectCategoryLinkListItemAction", this, false);
                this._preparedForAction = true;
            }
        }
    },
    exitDocument: {
        value: function () {
            if (this._preparedForAction) {
                this._preparedForAction = false;
                this.removeEventListener("selectCategoryLinkListItemAction", this);
            }
        }
    },

    tab1: {
        get: function() {
            return this.categories[0];
        }
    },

    createPanel: {
        value: function(index) {

            var self = this;
            return require.async("ui/link/link-list.reel").then(function(exports) {
                var pane = new exports.LinkList();
                pane.identifier = "linkslist";
                pane.ownerComponent = self;
                pane.defineBinding("links", {"<-": "categories["+index+"].links", source: self});
                pane.defineBinding("title", {"<-": "categories["+index+"].name", source: self});
                return pane;
            });
        }
    },

    userAccountAdministration: {
        get: function () {
            return this.createPanel(0);
        }
    },

    pdcResources: {
        get: function () {
            return this.createPanel(1);
        }
    },

    hazardInformationResources: {
        get: function () {
            return this.createPanel(2);
        }
    }
});
