// contour-framework/ui/brief/brief-inspector.reel/brief-inspector.js
/**
 * @module ui/brief-inspector.reel
 */
var Component = require("montage/ui/component").Component;

/**
 * @class BriefInspector
 * @extends Component
 */
exports.BriefInspector = Component.specialize(/** @lends BriefInspector.prototype */ {
    _tabs: {
        value: []
    },

    _availableTabs: {
        value: null
    },

    _panels: {
        value: {}
    },

    availableTabs: {
        set: function (value) {
            this._availableTabs = value;
        },
        get: function () {
            return this._availableTabs || [];
        }
    },

    tabBar: {
        value: null
    },

    _briefDetails: {
        value: null
    },

    _needsToGenerateTabs: {value: false},

    briefDetails: {
        get: function () {
            return this._briefDetails;
        },
        set: function (value) {
            if (value && value !== this._briefDetails) {
                this._briefDetails = value;
                this._needsToGenerateTabs = true;
                this.needsDraw = true;
            }
        }
    },

    draw: {
        value: function () {
            if (this._needsToGenerateTabs) {
                this._needsToGenerateTabs = false;

                var tabs = this._briefDetails.brief.tabs,
                    keys = tabs.map(function (tab) {
                        return tab.name;
                    });

                this.availableTabs = this._tabs.filter(function (uiTab) {
                    var index = keys.indexOf(uiTab.key);
                    uiTab._index = index;

                    return index !== -1;
                });

                this.selectedIndex = 0;
            }
        }
    },

    situationalAwarnessResources: {
        get: function () {
            return this.getPanel(this._tabs[0]);
        }
    },

    regionalOverviewResources: {
        get: function () {
            return this.getPanel(this._tabs[1]);
        }
    },

    riskVulernabilityResources: {
        get: function () {
            return this.getPanel(this._tabs[2]);
        }
    },

    bioHealthResources: {
        get: function () {
            return this.getPanel(this._tabs[3]);
        }
    },

    historicalHazardsResources: {
        get: function () {
            return this.getPanel(this._tabs[4]);
        }
    },

    getPanel: {
        value: function (tab) {
            var self = this,
                panelName = tab.paneReference.propertyName;

            if (!this._panels[panelName]) {
                this._panels[panelName] = require.async("ui/brief/brief-inspector-details.reel").then(function (exports) {
                    var pane = new exports.BriefInspectorDetails();
                    pane.identifier = "briefInspectorDetails_" + tab.id;
                    pane.ownerComponent = self;
                    pane.tab = tab;
                    pane.defineBinding("briefDetails", {
                        "<-": "_briefDetails",
                        source: self
                    });
                    return pane;
                });
            }
            return this._panels[panelName];
        }
    }

});
