// contour-framework/ui/drawing/figure-image-settings.reel/figure-image-settings.js
/**
 * @module ui/drawing/figure-image-settings.reel
 */
var Component = require("montage/ui/component").Component;

/**
 * @class FigureImageSettings
 * @extends Component
 */
exports.FigureImageSettings = Component.specialize(/** @lends FigureImageSettings# */ {
    constructor: {
        value: function FigureImageSettings() {

        }
    },
    enterDocument: {
        value: function(firstTime) {
            if (!firstTime) {
                this.prepareForActivationEvents();
            }
        }
    },
    exitDocument: {
        value: function() {
            if (this._preparedForAction) {
                this._preparedForAction = false;
                this.templateObjects.resetAspectRatio.removeEventListener("action",this);
            }
        }
    },
    _opacity: {
        value: 1
    },
    opacity: {
        set: function (value) {
            if (this._opacity != value) {
                this._opacity = value;
                this.needsDraw = true;
            }
        },
        get: function () {
            return this._opacity;
        }
    },
    _preserveAspectRatio: {
        value: false
    },
    preserveAspectRatio: {
        get: function() {
            return this._preserveAspectRatio;
        },
        set: function(value) {
            this._preserveAspectRatio = value;
        }
    },
    hasResetAspectRatio: {
        value: null
    },
    _preparedForAction: {
        value: false
    },
    prepareForActivationEvents: {
        value: function() {
            if (!this._preparedForAction) {
                this._preparedForAction = true;
                this.templateObjects.resetAspectRatio.addEventListener("action",this,false);
            }
        }
    },
    handleAction: {
        value: function(e) {
            e.stopPropagation();
            this.dispatchEventNamed("resetAspectRatioAction",false,false, {});
        }
    }
});
