// contour-framework/ui/control/floating-button.reel/floating-button.js
/**
 * @module "ui/mobile-contextual-menu.reel"
 * @requires montage/ui/component
 */
var Component = require("montage/ui/component").Component;

/**
 * @class MobileContextualMenu
 * @extends Component
 */
exports.FloatingButton = Component.specialize(/** @lends MobileContextualMenu.prototype */ {

    HOME: {
        value: {
            "class": "contour-FloatingButton--home",
            "action": "mainMenuShowAction"
        },
        writeable: false
    },

    MENU: {
        value: {
            "class": "contour-FloatingButton--menu",
            "action": "mainMenuHideAction"
        },
        writeable: false
    },

    PANEL: {
        value: {
            "class": "contour-FloatingButton--panel",
            "action": "contextMenuShowAction"
        },
        writeable: false
    },

    _context: {
        value: null
    },

    context: {
        get: function () {
            return this._context || this.HOME;
        },
        set: function (context) {
            if (context != this._context) {
                this.classList.remove((this._context || this.HOME).class);
                this._context = context;
                this.classList.add(this._context.class);
            }
        }
    },

    prepareForActivationEvents: {
        value: function () {
            this.addEventListener("action", this, false);
        }
    },

    handleButtonAction: {
        value: function (event) {
            event.stopPropagation();
            this.dispatchEventNamed(this.context.action, true, true);
        }
    }

});
