// contour-framework/ui/identify/identify-feature-table.info/sample/ui/main.reel/main.js
/**
 * @module ui/main.reel
 */
var Component = require("montage/ui/component").Component,
    HazardSelector = require("contour-framework/logic/mock-service/hazard-selector").HazardSelector,
    HazardService = require("contour-framework/logic/mock-service/hazard-service").HazardService;

/**
 * @class Main
 * @extends Component
 */
exports.Main = Component.specialize(/** @lends Main# */ {
    constructor: {
        value: function Main() {

            var service = new HazardService(),
                selector = new HazardSelector(),
                self = this;
            service.query(selector).then(function (data) {
               var features = data;
                self.features = features.slice(0, 15).map(function (value) {
                    value.properties.categoryId = "EVENT";
                    return value;
                });
            });
        }
    },
    features: {
        value: []
    },
    fields: {
        value: [
            {
                propertyName: "name",
                alias: "Name",
                type: "string",
                formattedValue: function(value) {
                    return value;
                }
            },
            {
                propertyName: "creationDate",
                alias: "Create Date",
                type: "date",
                formattedValue: function(value) {
                    return value.format("M.DD.YYYY H\\h:mm\\m");
                }
            },
            {
                propertyName: "typeId",
                alias: "Type",
                type: "string",
                formattedValue: function(value) {
                    return value;
                }
            },
            {
                propertyName: "categoryId",
                alias: "Category",
                type: "string",
                formattedValue: function(value) {
                    return value;
                }
            }
        ]
    }

});
