// contour-framework/ui/control/tab-bar.info/sample/ui/pane.reel/pane.js
var Component = require("montage/ui/component").Component;

/**
 * @class Pane
 * @extends external:Component
 */
exports.Pane = Component.specialize(/** @lends Pane# */ {

    succession: {
        value: undefined
    },

    name: {
        value: undefined
    },

    hasToggle: {
        value: false
    },

    substitue: { // For pane A, this will be set to pane B when it is selected.
        value: undefined
    },

    handleToggleAction: {
        value: function (event) {
            var self = this;
            if (this._paneA) {
                // This is pane B, switch to pane A.
                this.succession.content = this._paneA.substitute = this._paneA;
            } else if (this._paneB) {
                // This is pane A and pane B has been loaded, switch to pane B.
                this.succession.content = this.substitute = this._paneB;
            } else if (this._paneB !== null) {
                // This is pane A and pane B has not yet been loaded, load pane
                // B and then switch to it, making sure it is only loaded once.
                this._paneB = null; // Marker to prevent other load attempts.
                require.async("ui/pane.reel").then(function(exports) {
                    console.log("Creating pane " + self.name.slice(0, -1) + "B");
                    var pane = new exports.Pane();
                    pane.name = self.name.slice(0, -1) + "B";
                    pane.hasToggle = true;
                    pane.succession = self.succession;
                    pane._paneA = self;
                    self._paneB = pane;
                    self.handleToggleAction();
                });
            }
        }
    }

});
