// contour-framework/ui/social/social-video.reel/social-video.js
/**
 * @module ui/social-video.reel
 */
var Component = require("montage/ui/component").Component;

/**
 * @class SocialVideo
 * @extends Component
 */
exports.SocialVideo = Component.specialize(/** @lends SocialVideo# */ {

    constructor: {
        value: function SocialVideo() {

        }
    },

    _currentPlayingItem: {
        value: undefined
    },


    _preparedForAction: { value: false },

    _settingsPane: {value: null},

    enterDocument: {
        value: function (firstTime) {
            if (!firstTime) {
                this.prepareForActivationEvents();
            }
        }
    },

    exitDocument: {
        value: function () {
            if (this._preparedForAction) {
                this._preparedForAction = false;
                this._titleBar.removeEventListener("settingsAction", this, false);
            }
        }
    },

    handleSettingsAction: {
        value: function (e) {
            e.stopPropagation();
            this.settingsPane.succession = this.succession;
            this.succession.components.push(this.settingsPane);
        }
    },

    playVideo: {
        value: function (component) {
            if (component && component !== this._currentPlayingItem) {
                var src = "https://www.youtube.com/embed/" + component.tweet.youtube_code;
                if (this._currentPlayingItem) {
                    this._currentPlayingItem.player.element.setAttribute("src", "");
                    this._currentPlayingItem.displayVideoPlayer = false;
                }
                component.player.element.setAttribute("src", src);
                component.displayVideoPlayer = true;
                this._currentPlayingItem = component;
            }
        }
    },

    prepareForActivationEvents: {
        value: function () {
            if (!this._preparedForAction) {
                this._preparedForAction = true;
                this._titleBar.addEventListener("settingsAction", this, false);
            }

        }
    },

    settingsPane: { value: null },

    succession: { value: null }

    //handleAction: {
    //    value: function () {
    //        this.dispatchEventNamed("tweetsNextPage", true, true, null);
    //    }
    //}

});
