// contour-framework/ui/control/select-bar.info/sample/ui/main.reel/main.js
/**
 * @module ui/main.reel
 * @requires montage/ui/component
 */
var Component = require("montage/ui/component").Component;

/**
 * @class Main
 * @extends Component
 */
exports.Main = Component.specialize(/** @lends Main.prototype */ {

    constructor: {
        value: function Main() {

            this.dynamicOptions1 = this._options1;
            this.dynamicOptions2 = this._options1;
        }
    },

    _isPrepared: {
        value: false
    },

    prepareForActivationEvents: {
        value: function () {
            if (!this._isPrepared) {
                this.addEventListener("headingAction", this, false);
                this.addEventListener("select", this, false);
                this._isPrepared = true;
            }
        }
    },

    handleHeadingAction: {
        value: function (event) {
            var i, n, console, select, heading = event.detail.headingBar;
            if (heading.identifier == "heading6") {
                this.dynamicOptions1 = this._addOptions1ToSelect6 ?  this._options1.slice() : this._options2.slice();
                this._addOptions1ToSelect6 = !this._addOptions1ToSelect6;
            } else if (heading.identifier == "heading7") {
                this.dynamicOptions2 = this._addOptions1ToSelect7 ? this._options1 : this._options2;
                this._addOptions1ToSelect7 = !this._addOptions1ToSelect7;
            } else {
                for (i = 0, n = heading.selects.length; i < n; ++i) {
                    select = this.templateObjects[heading.selects[i]];
                    console = this.templateObjects[select.console];
                    console.prefix = "Reset to ";
                    if (select.defaults) {
                        select.values = select.defaults;
                    } else {
                        select.value = select.default;
                    }
                }
            }
        }
    },

    handleSelect: {
        value: function (event) {
            var select = event.detail.selectBar,
                console = this.templateObjects[select.console];
            console.prefix = "Set to ";
        }
    },
    dynamicOptions1: {
        value: []
    },
    dynamicOptions2: {
        value: []
    },
    _addOptions1ToSelect6: {
        value: false
    },
    _addOptions1ToSelect7: {
        value: false
    },
    _options1: {
        value: [{
            label:"A",
            value:"A"
        },
        {
            label:"B",
            value:"B"
        }
        ]
    },
    _options2: {
        value: [
            {
                label:"One",
                value:"1"
            },
            {
                label:"Two",
                value:"2"
            },
            {
                label:"Three",
                value:"3"
            }
        ]
    }

});
