// contour-framework/ui/hazard/hazard-type-editor.reel/hazard-type-editor.js
var Component = require("montage/ui/component").Component,
    HazardTypeSelector = require("logic/mock-service/hazard-type-selector").HazardTypeSelector;

/**
 * @class HazardTypeEditor
 * @extends Component
 */
exports.HazardTypeEditor = Component.specialize(/** @lends HazardTypeEditor.prototype */{
    selection: {value: []},

    multiSelect: {value: true},

    multiLine: {value: true},

    _options: {
        value: null
    },

    options: {
        get: function () {
            if (!this._options) {
                this._options = this.application.delegate.layers.hazard.hazardTypes;
            }
            return this._options;
        }
    },

    templateDidLoad: {
        value: function () {
            var self = this;
            self.needsDraw = true;
            self.templateObjects.headingBar.needsDraw = true;
            var type;
            for (var i = 0; i < self.options.length; i++) {
                type = self.options[i];
                type.icon = window.localStorage[type.id];
            }
        }
    }
});
