// contour-framework/ui/control/error-bar.reel/error-bar.js
/**
 * @module ui/control/error-bar.reel
 * @requires montage/ui/component
 */
var Component = require("montage/ui/component").Component;

/**
 * @class ErrorBar
 * @extends Component
 */
exports.ErrorBar = Component.specialize(/** @lends ErrorBar */{
    _isVisible: {
        value: false
    },

    isVisible: {
        set: function (isVisible) {
            clearTimeout(this._hideTimeout);

            if (isVisible) {
                this.show();
            } else {
                this.hide();
            }

            this.needsDraw = true;
        },
        get: function () {
            return this._isVisible;
        }
    },

    timeout: {
        value: 5000
    },

    _message: {
        value: ""
    },

    _hideTimeout: {
        value: null
    },

    message: {
        set: function (message) {
            this._message = message;
            this.isVisible = !!message;
        },
        get: function () {
            return this._message;
        }
    },

    constructor: {
        value: function () {

        }
    },

    show: {
        value: function () {
            this._isVisible = true;

            this.element.classList.remove("contour-ErrorBar--hidden");

            if (this.timeout) {
                this._hideTimeout = setTimeout(this.hide.bind(this), this.timeout);
            }
        }
    },

    hide: {
        value: function () {
            this._isVisible = false;

            this.element.classList.add("contour-ErrorBar--hidden");
        }
    },

    handleHideButtonAction: {
        value: function (event) {
            event.stopPropagation();
            this.isVisible = false;
            this.dispatchEventNamed("hideAction", true, true);
        }
    }
});
