// contour-framework/logic/model/bounds.js
var Montage = require("montage").Montage,
    ObjectDescriptor = require("montage-data/logic/model/object-descriptor").ObjectDescriptor,
    ProjectionUtilities = require("logic/map/projection-utilities").ProjectionUtilities;

/**
 * Implementation independent representation of bounds.
 *
 * @class Bounds
 * @extends external:Montage
 */
var Bounds = exports.Bounds = Montage.specialize(/** @lends Bounds# */ {

    xMin: {
        value: undefined
    },

    xMax: {
        value: undefined
    },

    yMin: {
        value: undefined
    },

    yMax: {
        value: undefined
    },

    equals: {
        value: function (other) {
            return other === this ||
                   other && other.xMin === this.xMin &&
                            other.yMin === this.yMin &&
                            other.xMax === this.xMax &&
                            other.yMax === this.yMax;
        }
    },

    splitAlongAntimeridian: {
        value: function () {
            if (!this._split) {
                this._split = [];
                if (this.xMin <= this.xMax) {
                    this._split.push(this);
                } else {
                    this._split.push(Bounds.withCoordinates(this.xMin, this.yMin, 179.99999, this.yMax));
                    this._split.push(Bounds.withCoordinates(-179.99999, this.yMin, this.xMax, this.yMax));
                }
            }
            return this._split;
        }
    }

}, {

    withCoordinates: {
        value: function(xMin, yMin, xMax, yMax, projection) {
            var bounds = new Bounds(),
                minimums = projection ? projection.inverseProjectPoint([xMin, yMin]) : [xMin, yMin],
                maximums = projection ? projection.inverseProjectPoint([xMax, yMax]) : [xMax, yMax];
            bounds.xMin = minimums[0];
            bounds.yMin = minimums[1];
            bounds.xMax = maximums[0];
            bounds.yMax = maximums[1];
            return bounds;
        }
    }

});

Bounds.EARTH = Bounds.withCoordinates(-180.0, -85.05112878, 180.0, 85.05112878);
