// contour-framework/logic/model/hazard-layer.js
var Layer = require("logic/model/layer").Layer,
    Feature = require("logic/model/feature").Feature,
    GeometryType = require("logic/model/geometry-type").GeometryType,
    Hazard = require("logic/model/hazard").Hazard,
    HazardIconRenderer = require("logic/map/hazard-icon-renderer").HazardIconRenderer,
    ObjectDescriptor = require("montage-data/logic/model/object-descriptor").ObjectDescriptor;

/**
 * A layer of hazards.
 *
 * @class
 * @extends external:Montage
 */
exports.HazardLayer = Layer.specialize(/** @lends HazardLayer.prototype */ {

    id: {
        value: 8
    },

    name: {
        value: "PDC Active Hazards"
    },

    type: {
        value: Layer.Type.FEATURE
    },

    geometryType: {
        value: GeometryType.POINT
    },

    depth: {
        value: 5000
    },

    areFeaturesGlobal: {
        get: function () {
            return true;
        }
    },

    isClustering: {
        value: true
    },

    renderer: {
        get: function () {
            if (!this._renderer) {
                this._renderer = new HazardIconRenderer();
            }
            return this._renderer;
        }
    },

    /**
     * @type {Array.<Hazard.Category>}
     */
    hazardCategories: {
        get: function () {
            if (!this._hazardCategories) {
                this._hazardCategories = [];
            }
            return this._hazardCategories;
        }
    },

    /**
     * @type {Array.<Hazard.Severity>}
     */
    hazardSeverities: {
        get: function () {
            if (!this._hazardSeverities) {
                this._hazardSeverities = [];
            }
            return this._hazardSeverities;
        }
    },

    /**
     * @type {Array.<Hazard.Type>}
     */
    hazardTypes: {
        get: function () {
            if (!this._hazardTypes) {
                this._hazardTypes = [];
            }
            return this._hazardTypes;
        }
    }

    // Other properties are inherited from Layer.

}, /** @lends HazardLayer */ {

    /** 
     * @type {external:ObjectDescriptor}
     */
    TYPE: {
        get: ObjectDescriptor.getterFor(exports, "HazardLayer", {
            "_features_all": {destinationType: Hazard.TYPE},
            "_features_visible": {destinationType: Hazard.TYPE},
            "_featureFields_all": {destinationType: null},
            "_featureFields_detail": {destinationType: null},
            "_featureFields_summary": {destinationType: null},
            "hazardCategories": {destinationType: null},
            "hazardSeverities": {destinationType: null},
            "hazardTypes": {destinationType: null},
            "hazards": {destinationType: Hazard.TYPE},
            "url": {destinationType: null}
        })
    }

});
