// contour-framework/logic/model/map-service.js
var Montage = require("montage").Montage,
    ObjectDescriptor = require("montage-data/logic/model/object-descriptor").ObjectDescriptor,
    Protocol = require("logic/model/protocol").Protocol;

/**
 * Responsible for providing information about and data for a set of layers.
 *
 * @class MapService
 * @extends external:Montage
 */
exports.MapService = Montage.specialize(/** @lends MapService# */ {

    url: {
        value: undefined
    },

    protocol: {
        get: function () {
            return this._protocol;
        },
        set: function (protocol) {
            this._protocol = protocol instanceof Protocol ? protocol :
                             protocol ?                     Protocol.forId(protocol) :
                                                            null;
        }
    },

    protocolVersion: {
        value: undefined
    },

    softwareName: {
        value: undefined
    },

    layers: {
        get: function () {
            if (!this._layers) {
                this._layers = {};
            }
            return this._layers;
        }
    }

}, {

    /**
     * The Montage Data type of map services.
     *
     * @type {external:ObjectDescriptor}
     */
    TYPE: {
        get: ObjectDescriptor.getterFor(exports, "MapService")
    }

});
