// contour-framework/ui/control/edit-bar.info/sample/ui/main.reel/main.js
/**
 * @module ui/main.reel
 * @requires montage/ui/component
 */
var Component = require("montage/ui/component").Component;

/**
 * @class Main
 * @extends Component
 */
exports.Main = Component.specialize(/** @lends Main.prototype */ {
    constructor: {
        value: function Main() {

        }
    },

    handleEditBar1CancelAction: {
        value: function (event) {
            event.stopPropagation();
            this.templateObjects.editBar1Description.pushAction("cancel");
        }
    },

    handleEditBar1CreateAction: {
        value: function (event) {
            event.stopPropagation();
            this.templateObjects.editBar1Description.pushAction("create");
        }
    },

    handleEditBar2CancelAction: {
        value: function (event) {
            event.stopPropagation();
            this.templateObjects.editBar2Description.pushAction("cancel");
        }
    },

    handleEditBar2ImportAction: {
        value: function (event) {
            event.stopPropagation();
            this.templateObjects.editBar2Description.pushAction("import");
        }
    },

    handleEditBar3CancelAction: {
        value: function (event) {
            event.stopPropagation();
            this.templateObjects.editBar3Description.pushAction("cancel");
        }
    },

    handleEditBar3SaveAction: {
        value: function (event) {
            event.stopPropagation();
            this.templateObjects.editBar3Description.pushAction("save");
        }
    },

    handleEditBar4CancelAction: {
        value: function (event) {
            event.stopPropagation();
            this.templateObjects.editBar4Description.pushAction("cancel");
        }
    },

    handleEditBar4SaveAction: {
        value: function (event) {
            event.stopPropagation();
            this.templateObjects.editBar4Description.pushAction("save");
        }
    }
});
