// contour-framework/ui/social/social-twitter.reel/social-twitter.js
/**
 * @module ui/social-twitter.reel
 */
var Component = require("montage/ui/component").Component;

/**
 * @class SocialTwitter
 * @extends Component
 */
exports.SocialTwitter = Component.specialize(/** @lends SocialTwitter.prototype */ {

    //enterDocument: {
    //    value: function (firstDraw) {
    //        if (firstDraw) {
    //            this.element.addEventListener('scroll', this);
    //        }
    //    }
    //},

    _preparedForAction: { value: false },



    enterDocument: {
        value: function (firstTime) {
            if (!firstTime) {
                this.prepareForActivationEvents();
            }
        }
    },

    exitDocument: {
        value: function () {
            if (this._preparedForAction) {
                this._preparedForAction = false;
                this._titleBar.removeEventListener("settingsAction", this, false);
            }
        }
    },

    handleSettingsAction: {
        value: function (e) {
            e.stopPropagation();
            this.settingsPane.succession = this.succession;
            this.succession.components.push(this.settingsPane);

        }
    },

    layer: {
        value: null
    },

    prepareForActivationEvents: {
        value: function () {
            if (!this._preparedForAction) {
                this._preparedForAction = true;
                this._titleBar.addEventListener("settingsAction", this, false);

            }

        }
    },

    settingsPane: { value: null },

    succession: { value: null }

    //_lastNextPageScrollHeight: {
    //    value: 0
    //},

    //
    //handleAction: {
    //    value: function () {
    //        this.dispatchEventNamed("tweetsNextPage", true, true, null);
    //    }
    //},
    //
    //handleScroll: {
    //    value: function (event) {
    //        if (this.hasMoreData &&
    //            this._lastNextPageScrollHeight != this.element.scrollHeight &&
    //            this.element.scrollHeight - this.element.scrollTop - 30 < this.element.clientHeight
    //        ) {
    //            this._lastNextPageScrollHeight = this.element.scrollHeight;
    //            this.dispatchEventNamed("tweetsNextPage", true, true, null);
    //        }
    //    }
    //}
});
