// contour-framework/ui/camera/camera-list-settings.reel/camera-list-settings.js
/**
 * @module ui/camera/camera-list-settings.reel
 * @requires montage/ui/component
 */
var Component = require("montage/ui/component").Component,
    moment = require("moment-timezone");

/**
 * @class CameraListSettings
 * @extends Component
 */
exports.CameraListSettings = Component.specialize(/** @lends CameraListSettings# */ {
    constructor: {
        value: function CameraListSettings() {

        }
    },

    _selector: {
        value: null
    },

    selector: {
        set: function (value) {
            this._selector = value;

        },
        get: function () {
            return this._selector;
        }
    },

    _preparedForAction: {
        value: false
    },

    enterDocument: {
        value: function (firstTime) {
            if (this._selector && this.templateObjects) {
                this._transferFilterFromSelectorToForm();
            }
        }
    },

    exitDocument: {
        value: function () {
            if (this._preparedForAction) {
                this._preparedForAction = false;
                this.templateObjects.sortLegendBar.removeEventListener("headingAction", this);
            }
        }
    },

    prepareForActivationEvents: {
        value: function () {
            if (!this._preparedForAction) {
                this._preparedForAction = true;
                this.templateObjects.sortLegendBar.addEventListener("headingAction", this, false);
            }
        }
    },

    handleBackAction: {
        value: function (event) {
            event.stopPropagation();
            this._transferFilterFromFormToSelector();
            this.dispatchEventNamed("cameraListSettingsBackAction", true, true, {});
        }
    },

    handleSortLegendBarHeadingAction: {
        value: function (event) {
            event.stopPropagation();
            this.templateObjects.sortFieldSelectBar.value = this.templateObjects.sortFieldSelectBar.default;
            this.templateObjects.sortDirectionSelectBar.value = this.templateObjects.sortDirectionSelectBar.default;
        }
    },

    _transferFilterFromFormToSelector: {
        value: function () {
            //this.selector.filter.sort.field = this.templateObjects.sortFieldSelectBar.value;
            //this.selector.filter.sort.direction = this.templateObjects.sortDirectionSelectBar.value;
        }
    },

    _transferFilterFromSelectorToForm: {
        value: function () {
            var templateObjects = this.templateObjects, filter = this._selector.filter;
            templateObjects.sortFieldSelectBar.value = filter.sort.field ? filter.sort.field : templateObjects.sortFieldSelectBar.default;
            templateObjects.sortDirectionSelectBar.value = filter.sort.direction ? filter.sort.direction : templateObjects.sortDirectionSelectBar.default;
        }
    }
});
