// contour-framework/ui/product/product-summary.reel/product-summary.js
/**
 * @module ui/feature-summary.reel
 * @requires montage/ui/component
 */
var Component = require("montage/ui/component").Component;

/**
 * @class FeatureSummary
 * @extends Component
 */
exports.ProductSummary = Component.specialize(/** @lends FeatureSummary# */ {

    constructor: {
        value: function FeatureSummary() {

        }
    },

    _feature: {
        value: null
    },

    _layer: {
        value: null
    },

    _product: {
        value: null
    },

    displayProductName: {
        value: true
    },

    data: {
        value: []
    },

    draw: {
        value: function() {
            if ( this.feature && this.icon ) {
                var style = this.feature.layer.renderer.render(this.feature.properties);
                this.icon.style.backgroundImage = "url('" + style.url + "')";
            }
        }
    },

    feature: {
        get: function() {
            return this._feature;
        },
        set: function( value ) {
            if ( value && value !== this._feature ) {
                this._feature = value;
                this.product = value.properties.data;
                this.needsDraw = true;
            }
        }
    },

    product: {
        set: function( product ) {
            var coordinates;
            var date = product.timelineDate;
            if ( this._product !== product ) {
                coordinates = this.feature.geometry.coordinates;

                this._product = product;

                this.data = [{
                    label: "LAT",
                    value: coordinates[0].toFixed(3)
                }, {
                    label: "LNG",
                    value: coordinates[1].toFixed(3)
                }, {
                    label: "TIMELINE DATE",
                    value: ("0" + (date.getMonth() + 1)).substr(-2) + "." + ("0" + date.getDate()).substr(-2) + "." + date.getFullYear()
                }];
            }
        },
        get: function() {
            return this._product;
        }
    },

    icon: {
        value: null
    }

});
