// contour-framework/ui/layer/import/layer-import-list-item.reel/layer-import-list-item.js
/**
 * @module ui/layer/layer-import-list-item.reel
 * @requires montage/ui/component
 */
var Component = require("montage/ui/component").Component;

/**
 * @class LayerImportListItem
 * @extends Component
 */
exports.LayerImportListItem = Component.specialize(/** @lends LayerImportListItem# */ {
    constructor: {
        value: function LayerImportListItem() {

        }
    },

    layer: {
        value: null // Value to be set by this component's owner.
    },

    templateDidLoad: {
        value: function () {
            this.element.classList.add(this._getType(this.layer));
        }
    },

    prepareForActivationEvents: {
        value: function () {
            if (!this._preparedForAtion) {
                var self = this,
                    type = this.layer.protocol ? "service" : "file";

                this._preparedForAtion = true;

                this.element.addEventListener("click", function () {
                    self.dispatchEventNamed("itemSelected", true, true, {
                        type: type,
                        data: self.layer
                    });
                });
            }
        }
    },

    _getType: {
        value: function (layer) {
            if (layer.type) {
                return layer.type;
            } else if (layer.protocol === "WMS") {
                return "wms";
            } else {
                return "ags";
            }
        }
    }
});
