// contour-framework/ui/product/product-type-editor.reel/product-type-editor.js
/**
 * @module ui/product-type-editor.reel
 */
var Component = require("montage/ui/component").Component,
    Product = require("logic/model/product").Product;

/**
 * @class ProductTypeEditor
 * @extends Component
 */
exports.ProductTypeEditor = Component.specialize(/** @lends ProductTypeEditor# */ {
    selection: {value: []},

    multiSelect: {value: true},

    _options: { value: null },

    options: {
        get: function () {
            if (!this._options) {
                this._options = [
                    Product.Type.TEXT_PRODUCT,
                    Product.Type.FILE_PRODUCT,
                    Product.Type.URL_PRODUCT,
                    Product.Type.FOLDER_PRODUCT
                ];
            }
            return this._options;
        }
    },

    draw: {
        value: function () {
            // cannot add icon class to this.templateObjects.icons b/c the arg gets absorbed by repetition
            var typeComponent,
                iterations = this.templateObjects.selectionEditor.templateObjects.repetition.iterations;

            for (var i = 0; i < iterations.length; i++) {
                if (iterations[i]._childComponents) {
                    typeComponent = iterations[i]._childComponents[1];
                    typeComponent.classList.add("Icon-multi-"+typeComponent.data.name.toLowerCase());
                }
            }
        }
    }
});
