// contour-framework/ui/control/contour-succession.reel/contour-succession.js
/**
 * @module ui/control/contour-succession
 * @requires montage/ui/component
 */
var Succession = require("montage/ui/succession.reel").Succession;

/**
 * @class ContourSuccession
 * @extends Component
 */
exports.ContourSuccession = Succession.specialize(/** @lends ContourSuccession.prototype */{
    constructor: {
        value: function () {
            Succession.call(this);
        }
    },

    /**
     * Re-implement to not use push w/ animation
     */
    content: {
        get: function () {
            return this._content || (this._content = []);
        },
        set: function (value) {
            if (this.content && this.content.length) {
                this._prepareForBuild();
                this.domContent = null;
                this.content.length = 0;
            }

            if (value) {
                Succession.push.call(this, value);
            }
        }
    },

    push: {
        value: function (value) {
            this.contentBuildInCssClass = "contour-LegacyPanel-pane-drill-down-build-in";
            this.contentBuildInTransitionCssClass = null;
            this.contentBuildOutCssClass = "contour-LegacyPanel-pane-drill-down-build-out";

            Succession.push.call(this, value);

            this._resetSuccessionContentBuild();
        }
    },
    pop: {
        value: function () {
            this.contentBuildInCssClass = "contour-LegacyPanel-pane-back-build-in";
            this.contentBuildInTransitionCssClass = null;
            this.contentBuildOutCssClass = "contour-LegacyPanel-pane-back-build-out";

            Succession.pop.call(this);

            this._resetSuccessionContentBuild();
        }
    },
    _resetSuccessionContentBuild: {
        value: function () {
            this.contentBuildInCssClass = undefined;
            this.contentBuildInTransitionCssClass = undefined;
            this.contentBuildOutCssClass = undefined;
        }
    }
});
