// contour-framework/ui/product/product-settings.reel/product-settings.js
/**
 * @module "ui/product/product-settings.reel"
 * @requires montage/ui/component
 */
var Component = require("montage/ui/component").Component;

/**
 * @class ProductSettings
 * @extends Component
 */
exports.ProductSettings = Component.specialize(/** @lends ProductSettings.prototype */{

    _sortHeading: { value: null },

    _titleBar: { value: null },

    _viewHeading: { value: null },

    prepareForActivationEvents: {
        value: function () {
            this._sortHeading.addEventListener("headingAction", this, false);
            this._viewHeading.addEventListener("headingAction", this, false);
            this._titleBar.addEventListener("backAction", this, false);
        }
    },

    handleBackAction: {
        value: function (event) {
            event.stopPropagation();
            this.succession.components.pop();
        }
    },

    handleHeadingAction: {
        value: function (event) {
            event.stopPropagation();
            var i, n, select, heading = event.detail.headingBar;

            if (heading.identifier === "viewHeading") {
                this.templateObjects.displayToggle.checked = false;
            }

            for (i = 0, n = heading.selects.length; i < n; ++i) {
                select = this.templateObjects[heading.selects[i]];
                select.value = select.default;
            }
        }
    },

    settings: {value: null},

    succession: {value: null}
});
