// contour/ui/control/title-bar.reel/title-bar.js
/**
 * @module ui/control/title-bar.reel
 * @requires montage/ui/component
 */
var Component = require("montage/ui/component").Component;

/**
 * @class TitleBar
 * @extends Component
 */
exports.TitleBar = Component.specialize(/** @lends TitleBar.prototype */ {

    constructor: {
        value: function TitleBar() {
            this.super();
        }
    },

    /***************************************************************************
     * Properties
     */

    isLarge: {
        value: false
    },

    isSmall: {
        value: false
    },

    title: {
        value: null
    },

    hasBack: {
        value: false
    },

    hasDone: {
        value: false
    },

    hasClose: {
        value: false
    },

    hasSettings: {
        value: false
    },

    areSettingsHighlighted: {
        value: false
    },

    /***************************************************************************
     * Button events are handled here and converted to title bar events.
     */

    _preparedForAction: {
        value: false
    },

    prepareForActivationEvents: {
        value: function () {
            if (!this._preparedForAction) {
                this.addEventListener("action", this, false);
                this._preparedForAction = true;
            }
        }
    },

    handleBackAction: {
        value: function (event) {
            event.stopPropagation();
            this.dispatchEventNamed("backAction", true, true, {});
        }
    },

    handleCloseAction: {
        value: function (event) {
            event.stopPropagation();
            this.dispatchEventNamed("closeAction", true, true, {});
        }
    },

    handleSettingsAction: {
        value: function (event) {
            event.stopPropagation();
            this.dispatchEventNamed("settingsAction", true, true, {});
        }
    }

});
