// contour-framework/ui/location/location-inspector.reel/location-inspector.js
/**
 * @module ui/location/location-inspector.reel
 */
var Component = require("montage/ui/component").Component;

/**
 * @class LocationInspector
 * @extends Component
 */
exports.LocationInspector = Component.specialize(/** @lends LocationInspector# */ {
    errorText: {
        value: ""
    },

    constructor: {
        value: function LocationInspector() {

        }
    },

    _latitude: {
        value: 65
    },

    _longitude: {
        value: -146
    },

    __mgrsCoordinate: {
        value: "33UXP04"
    },

    _mgrsCoordinate: {
        get: function () {
            return this.__mgrsCoordinate;
        },
        set: function (value) {
            if (this.__mgrsCoordinate !== value) {
                this.__mgrsCoordinate = value;
            }
        }
    },

    _preparedForAction: {
        value: false
    },

    enterDocument: {
        value: function () {
            this.mapCenter = this.application.delegate.getCenter();
        }
    },

    handleAction: {
        value: function (e) {
            e.stopPropagation();
            this._moveMapToMgrsCoordinate();
        }
    },
    handleCoordinateAction: {
        value: function (e) {
            e.stopPropagation();
            if (this.__mapCenter && (this._longitude !== this.__mapCenter[0] || this._latitude !== this.__mapCenter[1])) {
                this.application.delegate.setCenter([this._longitude, this._latitude]);
            }
        }
    },

    handleHeadingAction: {
        value: function (e) {
            e.stopPropagation();
            this._moveMapToMgrsCoordinate();
        }
    },

    latitude: {
        get: function () {
            return this._latitude;
        },
        set: function (value) {
            if (this._latitude !== value) {
                this._latitude = value;
            }
        }
    },

    longitude: {
        get: function () {
            return this._longitude;
        },
        set: function (value) {
            if (this._longitude !== value) {
                this._longitude = value;
            }
        }
    },

    __mapCenter: {
        value: null
    },

    _mapCenter: {
        get: function () {
            return this.__mapCenter;
        },
        set: function (value) {
            if (value) {
                this.__mapCenter = value;
                this.longitude = value[0];
                this.latitude = value[1];
            }
        }
    },

    prepareForActivationEvents: {
        value: function () {
            if (!this._preparedForAction) {
                this.coordinateEditor.addEventListener("coordinateAction", this, false);
                this.mgrsHeading.addEventListener("headingAction", this, false);
                this.mgrsCoordinateEditor.addEventListener("action", this, false);
            }
        }
    },

    _moveMapToMgrsCoordinate: {
        value: function () {
            if (!this.application.delegate.setCenterWithMgrsCoordinates(this._mgrsCoordinate)) {
                this.error = new Error("Invalid MGRS format");
            } else {
                this.error = null;
            }
        }
    }

});
