// contour-framework/ui/hazard/hazard-severity-editor.reel/hazard-severity-editor.js
var Component = require("montage/ui/component").Component,
    HazardSeveritySelector = require("logic/mock-service/hazard-severity-selector").HazardSeveritySelector;

/**
 * @class HazardSeverityEditor
 * @extends Component
 */
exports.HazardSeverityEditor = Component.specialize(/** @lends HazardSeverityEditor.prototype */{
    selection: {value: []},

    multiSelect: {value: true},

    _options: {
        value: null
    },

    options: {
        get: function () {
            if (!this._options) {
                this._options = this.application.delegate.layers.hazard.hazardSeverities;
            }
            return this._options;
        }
    },

    draw: {
        value: function () {
            // cannot add icon class to this.templateObjects.icons b/c the arg gets absorbed by repetition
            var severity,
                iterations = this.templateObjects.selectionEditor.templateObjects.repetition.iterations;
            for (var i = 0; i < iterations.length; i++) {
                if (iterations[i]._childComponents) {
                    severity = iterations[i]._childComponents[1];
                    severity.classList.add("Icon-multi-"+severity.data.id.toLowerCase());
                }
            }
        }
    }

});
