// contour-framework/logic/mock-service/import-layer-service.js
/**
 * @module logic/layer-service.js
 */
var ObjectService = require("logic/mock-montage-data/object-service").ObjectService,
    ObjectStream = require("logic/mock-montage-data/object-stream").ObjectStream;

/**
 * @class LayerService
 * @extends ObjectService
 */
exports.ImportLayerService = ObjectService.specialize(/** @lends LayerService.prototype */ {
    _value: {
        value: []
    },
    constructor: {
        value: function LayerService() {
            this.super("import-layer");
        }
    },
    add: {
        value: function ( layer ) {
            this._value.push(layer);
        }
    },
    remove: {
        value: function ( layer ) {
            var index = this._value.indexOf(layer);
            if ( index !== -1 ) {
                return this._value.splice(index, 1);
            }
            return null;
        }
    },
    query: {
        value: function (selector) {
            var stream = new ObjectStream(selector);
            setTimeout(function ( ) {
                stream.resolve(this._value);
            }, 0);
            return stream;
        }
    }
});
