// contour-framework/logic/mock-service/import-layers-service.js
var DataService = require("montage-data/logic/service/data-service").DataService,
    Bindings = require("frb"),
    application = require("montage/core/application").application;

/**
 * Mock user layers service.
 *
 * @class
 * @extends external:DataService
 */
exports.ImportLayersService = DataService.specialize(/** @lends UserLayersService# */{}, {
    _folder: {
        value: null
    },
    _layers: {
        value: []
    },
    layers: {
        get: function () {
            return this._layers;
        }
    },
    addLayer: {
        value: function (file) {
            var importOverlay = application.map.importData(file.name, file.content, file.ext),
                layers = application.delegate.layers,
                layer = {
                    name: file.name,
                    overlay: importOverlay,
                    isSelectable: true
                };

            Bindings.defineBinding(layer, "isEnabled", {"<->": "overlay.enabled"});

            this.layers.push(layer);

            if (!this._folder) {
                this._folder = {
                    name: "Imported Layers",
                    children: this.layers,
                    parent: layers.tree
                };
                layers.tree.children.push(this._folder);
                layers.list.push(layer);
            }

            return layer;
        }
    },
    removeLayer: {
        value: function (layer) {
            var index = this.layers.indexOf(layer);
            if (index > -1) {
                this.layers.splice(index, 1);
            }
        }
    }
});
