// contour-framework/logic/mock-service/layer-service.js
/**
 * @module logic/layer-service.js
 */
var ObjectService = require("logic/mock-montage-data/object-service").ObjectService,
    ObjectStream = require("logic/mock-montage-data/object-stream").ObjectStream;

/**
 * @class LayerService
 * @extends ObjectService
 */
exports.LayerService = ObjectService.specialize(/** @lends LayerService.prototype */ {
    constructor: {
        value: function LayerService() {
            this.super("layer");
        }
    },

    _mockLayers: {
        value: [
            {
                title: "PDC Integrated Active Hazards",
                description: "Description for layer 1",
                color: "red",
                isEnabled: false,
                isHidden: false,
                opacity: 1,
                basicFields: [
                    {
                        key: "DATE_",
                        label: "Date"
                        //formatter: function () {}
                    },
                    {key: "DEPTH_KM", label: "Depth (km)"},
                    {key: "MAGNITUDE", label: "Magnitude"}
                ],
                extendedFields: [],
                filterFields: [
                    {key: "REGION", label:"Region", type: "string"},
                    {key: "INCIDENT_TIME", label:"Incident Time", type: "date"},
                    {key: "MAGNITUDE", label:"Magnitude", type: "number"}
                ]
            },
            {
                title: "Hazards and Events",
                description: "Description for layer 2",
                color: "blue",
                isEnabled: false,
                isHidden: false,
                opacity: 0.4,
                basicFields: [],
                extendedFields: [],
                filterFields: []
            },
            {
                title: "Observation and Forecasts",
                description: "Description for layer 3",
                color: "green",
                isEnabled: false,
                isHidden: false,
                opacity: 0.8,
                basicFields: [],
                extendedFields: [],
                filterFields: []
            },
            {
                title: "Risk and Vulnerability",
                description: "Description for layer 4",
                color: "gray",
                isEnabled: true,
                isHidden: false,
                opacity: 0.2,
                basicFields: [],
                extendedFields: [],
                filterFields: []
            },
            {
                title: "Base Map Data",
                description: "Description for layer 5",
                color: "gray",
                isEnabled: false,
                isHidden: false,
                opacity: 1,
                basicFields: [],
                extendedFields: [],
                filterFields: []
            },
            {
                title: "Estimated WInd Impacts (TAOS Model)",
                description: "Description for layer 5",
                color: "gray",
                isEnabled: false,
                isHidden: false,
                opacity: 1,
                basicFields: [],
                extendedFields: [],
                filterFields: []
            },
            {
                title: "Tsunami Watch and Warning Zones",
                description: "Description for layer 5",
                color: "gray",
                isEnabled: false,
                isHidden: false,
                opacity: 1,
                basicFields: [],
                extendedFields: [],
                filterFields: []
            },
            {
                title: "Base Map Data",
                description: "Description for layer 5",
                color: "gray",
                isEnabled: false,
                isHidden: false,
                opacity: 1,
                basicFields: [],
                extendedFields: [],
                filterFields: []
            },
            {
                title: "Base Map Data",
                description: "Description for layer 5",
                color: "gray",
                isEnabled: false,
                isHidden: false,
                opacity: 1,
                basicFields: [],
                extendedFields: [],
                filterFields: []
            },
            {
                title: "Base Map Data",
                description: "Description for layer 5",
                color: "gray",
                isEnabled: false,
                isHidden: false,
                opacity: 1,
                basicFields: [],
                extendedFields: [],
                filterFields: []
            },
            {
                title: "Base Map Data",
                description: "Description for layer 5",
                color: "gray",
                isEnabled: false,
                isHidden: false,
                opacity: 1,
                basicFields: [],
                extendedFields: [],
                filterFields: []
            },
            {
                title: "Base Map Data",
                description: "Description for layer 5",
                color: "gray",
                isEnabled: false,
                isHidden: false,
                opacity: 1,
                basicFields: [],
                extendedFields: [],
                filterFields: []
            },
            {
                title: "Base Map Data",
                description: "Description for layer 5",
                color: "gray",
                isEnabled: false,
                isHidden: false,
                opacity: 1,
                basicFields: [],
                extendedFields: [],
                filterFields: []
            },
            {
                title: "Base Map Data",
                description: "Description for layer 5",
                color: "gray",
                isEnabled: false,
                isHidden: false,
                opacity: 1,
                basicFields: [],
                extendedFields: [],
                filterFields: []
            },
            {
                title: "Base Map Data",
                description: "Description for layer 5",
                color: "gray",
                isEnabled: false,
                isHidden: false,
                opacity: 1,
                basicFields: [],
                extendedFields: [],
                filterFields: []
            }
        ]
    },

    query: {
        value: function (selector) {
            var self = this;
            var stream = new ObjectStream(selector);
            setTimeout(function ( ) {
                stream.add(self._mockLayers);
                stream.resolve(self._mockLayers);
            }, 0);
            return stream;
        }
    }
});
