// contour-framework/logic/mock-service/social-selector.js
/**
 * @module logic/social-selector.js
 */
var ObjectSelector = require("logic/mock-montage-data/object-selector").ObjectSelector;

/**
 * @class SocialSelector
 * @extends ObjectSelector
 *
 */
exports.SocialSelector = ObjectSelector.specialize(/** @lends TileSelector.prototype */ {

    constructor: {
        value: function SocialSelector(layer, extent, mediaType) { //}, toSysId, limit) {
            this.super("social");
            this.layer = layer;
            this.bounds = extent;
            this.mediaType = mediaType ? mediaType : "text";
            //this.toSysId = toSysId ? toSysId : false;
            //this.limit = limit ? limit : 25;
        }
    },

    _mediaType: {
        value: null
    },

    bounds: {
        value: null
    },

    cacheKeywords: {
        value: true
    },

    getKeywords: {
        value: false
    },

    layer: {
        value: undefined
    },

    limit: {
        value: undefined
    },

    mediaType: {
        get: function () {
            return this._mediaType;
        },
        set: function (value) {
            if (value && value !== this._mediaType) {
                this._mediaType = value;
                switch (this._mediaType) {
                    case "text":
                        this.limit = 250;
                        break;
                    case "image":
                        this.limit = 25;
                        break;
                    case "video":
                        this.limit = 10;
                        break;
                }
            }
        }
    }

});
