// contour-framework/logic/map/product-renderer.js
/**
 * @module logic/simple-renderer.js
 */
var Renderer = require("logic/map/renderer").Renderer;

/**
 * @class ProductRenderer
 * @extends Renderer
 *
 * This object is used by a import layer to convert geojson styles to
 * map engine styles
 */
var ProductRenderer = exports.ProductRenderer = Renderer.specialize(/** @lends ProductRenderer.prototype */ {

        constructor: {
            value: function ProductRenderer() {
            }
        },

        render: {
            value: function (properties) {
                var style = properties.style;

                if ( !style ) {
                    switch ( properties.type ) {
                        case "Point":
                            style = this._renderPoint(properties.data);
                            break;
                        case "LineString":
                            style = this._renderLine();
                            break;
                    }
                }

                return style;
            }
        },

        _renderPoint: {
            value: function (feature) {
                var url = L.Icon.Default.imagePath + "/geocoded_marker_icon_" + feature.type + ".png";
                var isRetina = false;
                var size = {
                    width: 45,
                    height: 45
                };
                var scale = size;

                if ( isRetina ) {
                    url = L.Icon.Default.imagePath + "/geocoded_marker_icon_" + feature.type + "-2x.png";
                    scale = {
                        width: 2 * size.width,
                        height: 2 * size.height
                    }
                }

                return {
                    url: url,
                    size: size,
                    scale: scale,
                    anchor: {
                        x: 23,
                        y: 23
                    }
                };
            }
        },

        _renderLine: {
            value: function ( ) {
                return {
                    strokeColor: "#0276D9",
                    strokeWeight: 5
                };
            }
        }
    });
