// contour-framework/ui/hazard/hazard-category-editor.reel/hazard-category-editor.js
var Component = require("montage/ui/component").Component,
    HazardCategorySelector = require("logic/mock-service/hazard-category-selector").HazardCategorySelector;

/**
 * @class HazardCategoryEditor
 * @extends Component
 */
exports.HazardCategoryEditor = Component.specialize(/** @lends HazardCategoryEditor.prototype */{
    selection: {value: []},

    multiSelect: {value: true},

    _options: {
        value: null
    },

    options: {
        get: function () {
            if (!this._options) {
                this._options = this.application.delegate.layers.hazard.hazardCategories;
            }
            return this._options;
        }
    },

    draw: {
        value: function () {
            // cannot add icon class to this.templateObjects.icons b/c the arg gets absorbed by repetition
            var category,
                iterations = this.templateObjects.selectionEditor.templateObjects.repetition.iterations;

            for (var i = 0; i < iterations.length; i++) {
                if (iterations[i]._childComponents) {
                    category = iterations[i]._childComponents[1];
                    category.classList.add("Icon-multi-"+category.data.id.toLowerCase());
                }
            }
        }
    }
});
