// contour-framework/ui/drawing/figure-placemark-selector.reel/figure-placemark-selector.js
/**
 * @module ui/drawing/figure-placemark-selector.reel
 */
var Component = require("montage/ui/component").Component;

/**
 * @class FigurePlacemarkSelector
 * @extends Component
 */
exports.FigurePlacemarkSelector = Component.specialize(/** @lends FigurePlacemarkSelector# */ {
    constructor: {
        value: function FigurePlacemarkSelector() {

        }
    },
    _figure: {
        value: null
    },
    figure: {
        set: function(value) {
            var i, length,marker;
            if (value && value !== this._figure) {
                for (i = 0, length = this.options.length; i < length; i++) {
                    marker = this.options[i];
                    if (marker.unicode == value.textCharacter) {
                        this.selectedMarker = marker;
                        break;
                    }
                }
                this._figure = value;
            }
        },
        get: function() {
            return this._figure;
        }
    },
    options: {
        value: [
            {
                id: "blank",
                icon: "Icon-placemark-blank",
                name: "Blank",
                unicode: "\ue65a"
            },
            {
                id: "circle",
                icon: "Icon-placemark-circle",
                name: "Circle",
                unicode: "\ue65b"
            },
            {
                id: "square",
                icon: "Icon-placemark-square",
                name: "Square",
                unicode: "\ue65c"
            },
            {
                id: "triangle",
                icon: "Icon-placemark-triangle",
                name: "Triangle",
                unicode: "\ue65d"
            },
            {
                id: "diamond",
                icon: "Icon-placemark-diamond",
                name: "Diamond",
                unicode: "\ue65e"
            },
            {
                id: "star",
                icon: "Icon-placemark-star",
                name: "Star",
                unicode: "\ue65f"
            },
            {
                id: "alpha-a",
                icon: "Icon-placemark-alpha-a",
                name: "A",
                unicode: "\ue660"
            },
            {
                id: "alpha-b",
                icon: "Icon-placemark-alpha-b",
                name: "B",
                unicode: "\ue661"
            },
            {
                id: "alpha-c",
                icon: "Icon-placemark-alpha-c",
                name: "C",
                unicode: "\ue662"
            },
            {
                id: "alpha-d",
                icon: "Icon-placemark-alpha-d",
                name: "D",
                unicode: "\ue663"
            },
            {
                id: "alpha-e",
                icon: "Icon-placemark-alpha-e",
                name: "E",
                unicode: "\ue664"
            },
            {
                id: "alpha-f",
                icon: "Icon-placemark-alpha-f",
                name: "F",
                unicode: "\ue665"
            },
            {
                id: "alpha-g",
                icon: "Icon-placemark-alpha-g",
                name: "G",
                unicode: "\ue666"
            },
            {
                id: "alpha-h",
                icon: "Icon-placemark-alpha-h",
                name: "H",
                unicode: "\ue667"
            },
            {
                id: "alpha-i",
                icon: "Icon-placemark-alpha-i",
                name: "I",
                unicode: "\ue668"
            },
            {
                id: "alpha-j",
                icon: "Icon-placemark-alpha-j",
                name: "J",
                unicode: "\ue669"
            },
            {
                id: "alpha-k",
                icon: "Icon-placemark-alpha-k",
                name: "K",
                unicode: "\ue66a"
            },
            {
                id: "alpha-l",
                icon: "Icon-placemark-alpha-l",
                name: "L",
                unicode: "\ue66b"
            },
            {
                id: "alpha-m",
                icon: "Icon-placemark-alpha-m",
                name: "M",
                unicode: "\ue66c"
            },
            {
                id: "alpha-n",
                icon: "Icon-placemark-alpha-n",
                name: "N",
                unicode: "\ue66d"
            },
            {
                id: "alpha-o",
                icon: "Icon-placemark-alpha-o",
                name: "O",
                unicode: "\ue66e"
            },
            {
                id: "alpha-p",
                icon: "Icon-placemark-alpha-p",
                name: "P",
                unicode: "\ue66f"
            },
            {
                id: "alpha-q",
                icon: "Icon-placemark-alpha-q",
                name: "Q",
                unicode: "\ue670"
            },
            {
                id: "alpha-r",
                icon: "Icon-placemark-alpha-r",
                name: "R",
                unicode: "\ue671"
            },
            {
                id: "alpha-s",
                icon: "Icon-placemark-alpha-s",
                name: "S",
                unicode: "\ue672"
            },
            {
                id: "alpha-t",
                icon: "Icon-placemark-alpha-t",
                name: "T",
                unicode: "\ue673"
            },
            {
                id: "alpha-u",
                icon: "Icon-placemark-alpha-u",
                name: "U",
                unicode: "\ue674"
            },
            {
                id: "alpha-v",
                icon: "Icon-placemark-alpha-v",
                name: "V",
                unicode: "\ue675"
            },
            {
                id: "alpha-w",
                icon: "Icon-placemark-alpha-w",
                name: "W",
                unicode: "\ue676"
            },
            {
                id: "alpha-x",
                icon: "Icon-placemark-alpha-x",
                name: "X",
                unicode: "\ue677"
            },
            {
                id: "alpha-y",
                icon: "Icon-placemark-alpha-y",
                name: "Y",
                unicode: "\ue678"
            },
            {
                id: "alpha-z",
                icon: "Icon-placemark-alpha-z",
                name: "Z",
                unicode: "\ue679"
            },
            {
                id: "num-one",
                icon: "Icon-placemark-num-one",
                name: "One",
                unicode: "\ue67a"
            },
            {
                id: "num-two",
                icon: "Icon-placemark-num-two",
                name: "Two",
                unicode: "\ue67b"
            },
            {
                id: "num-three",
                icon: "Icon-placemark-num-three",
                name: "Three",
                unicode: "\ue67c"
            },
            {
                id: "num-four",
                icon: "Icon-placemark-num-four",
                name: "Four",
                unicode: "\ue67d"
            },
            {
                id: "num-five",
                icon: "Icon-placemark-num-five",
                name: "Five",
                unicode: "\ue67e"
            },
            {
                id: "num-six",
                icon: "Icon-placemark-num-six",
                name: "Six",
                unicode: "\ue67f"
            },
            {
                id: "num-seven",
                icon: "Icon-placemark-num-seven",
                name: "Seven",
                unicode: "\ue680"
            },
            {
                id: "num-eight",
                icon: "Icon-placemark-num-eight",
                name: "Eight",
                unicode: "\ue681"
            },
            {
                id: "num-nine",
                icon: "Icon-placemark-num-nine",
                name: "Nine",
                unicode: "\ue682"
            },
            {
                id: "num-ten",
                icon: "Icon-placemark-num-ten",
                name: "Ten",
                unicode: "\ue683"
            }
        ]
    },
    optionsController: {
        set: function(value) {
            if (value) {
                value.select(this._selectedMarker);
            }
        }
    },
    _selectedMarker: {
        value: null
    },
    selectedMarker: {
        set: function(value) {
            if (value && this._selectedMarker !== value) {
                this._selectedMarker = value;
                if (this._figure && this._figure.textCharacter !== value.unicode) {
                    this._figure.textCharacter = value.unicode;
                }
            }
        },
        get: function() {
            return this._selectedMarker;
        }
    }
});
